### PR TITLE
Glue generate lambda set callers

### PR DESCRIPTION
**Pull request by folkertdev**
generates functions that evaluate roc closures, based on the layout of host-exposed functions. The mechanism for seeing whether a function is host-exposed is still based on the `as` aliases in platform `main.roc` modules. We'll have to figure out some other way of doing that eventually. 

the old mechanism also still fully works, the new functions are (for the moment) entirely unused.

### CODE REVIEW

File: crates/compiler/gen_llvm/src/llvm/build.rs [hunk 1]
```diff
@@ -3977,7 +3977,8 @@ fn expose_function_to_host_help_c_abi_v2<'a, 'ctx, 'env>(
             // In C, this is modelled as a function returning void
             (
                 &params[..],
-                &param_types[..param_types.len().saturating_sub(1)],
+                // &param_types[..param_types.len().saturating_sub(1)],
+                &param_types[..],
```
Comment (folkertdev) [inline crates/compiler/gen_llvm/src/llvm/build.rs:3981]: there is something subtler going on here. We'll surely revisit this when glue is more mature

File: crates/compiler/mono/src/ir.rs [hunk 2]
```diff
@@ -11129,3 +11189,343 @@ where
...
+pub struct GlueProc<'a> {
+    pub name: Symbol,
+    pub proc_layout: ProcLayout<'a>,
+    pub proc: Proc<'a>,
+}
+
+pub struct GlueProcs<'a> {
+    pub getters: Vec<'a, (Layout<'a>, Vec<'a, GlueProc<'a>>)>,
+    pub extern_names: Vec<'a, (LambdaSetId, RawFunctionLayout<'a>)>,
+}
```
Comment (folkertdev) [inline crates/compiler/mono/src/ir.rs:11208]: this code does more than just finding the function types which must be exposed. We ignore that other stuff for now though.

File: crates/compiler/gen_llvm/src/llvm/build.rs (hunk 1 above)
Comment (ayazhafiz) [reply crates/compiler/gen_llvm/src/llvm/build.rs:3981]: Maybe worth leaving a comment behind? Might be hard to find again.

File: crates/compiler/mono/src/ir.rs [hunk 3]
```diff
@@ -3057,36 +3064,129 @@ fn specialize_external_help<'a>(
...
+                            raw_function_layout = match raw_function_layout {
+                                RawFunctionLayout::Function(a, mut lambda_set, _) => {
+                                    lambda_set.ret = in_layout;
+                                    RawFunctionLayout::Function(a, lambda_set, in_layout)
+                                }
+                                RawFunctionLayout::ZeroArgumentThunk(x) => {
+                                    RawFunctionLayout::ZeroArgumentThunk(x)
+                                }
+                            };
+                        }
```
Comment (ayazhafiz) [inline crates/compiler/mono/src/ir.rs:3125]: is this needed?

**Review by ayazhafiz — APPROVED** [review]
Nice. Pretty great that the getters can happen at the IR level, Folkert.

File: crates/compiler/gen_llvm/src/llvm/build.rs (hunk 1 above)
Comment (folkertdev) [reply crates/compiler/gen_llvm/src/llvm/build.rs:3981]: no there is an assert right below

File: crates/compiler/mono/src/ir.rs (hunk 3 above)
Comment (folkertdev) [reply crates/compiler/mono/src/ir.rs:3125]: we'll see. We ran into a problem on a branch, but that did not have the latest type checker/ layout fixes.

**Pull request merged**